File: cpp/question1_bellman_ford.cpp
/**
 * Question 1 -- Bellman-Ford: Cursed Dungeon Routes (20 pts)
 */
#include <algorithm>
#include <string>
#include <vector>

const long long INF = 1000000000000000000LL;

struct Edge {
    int from;
    int to;
    int weight;
};

std::vector<long long> shortestDistances(const std::vector<Edge>& edges, int numNodes, int start) {
    std::vector<long long> dist(numNodes, INF);
    if (start < 0 || start >= numNodes) {
        return dist;
    }
    dist[start] = 0;
    for (int pass = 0; pass < numNodes - 1; pass++) {
        bool changed = false;
        for (const Edge& edge : edges) {
            if (edge.from < 0 || edge.from >= numNodes || edge.to < 0 || edge.to >= numNodes) {
                continue;
            }
            if (dist[edge.from] != INF && dist[edge.from] + edge.weight < dist[edge.to]) {
                dist[edge.to] = dist[edge.from] + edge.weight;
                changed = true;
            }
        }
        if (!changed) {
            break;
        }
    }
    return dist;
}

bool hasNegativeCycle(const std::vector<Edge>& edges, int numNodes, int start) {
    std::vector<long long> dist = shortestDistances(edges, numNodes, start);
    for (const Edge& edge : edges) {
        if (edge.from < 0 || edge.from >= numNodes || edge.to < 0 || edge.to >= numNodes) {
            continue;
        }
        if (dist[edge.from] != INF && dist[edge.from] + edge.weight < dist[edge.to]) {
            return true;
        }
    }
    return false;
}

std::vector<int> shortestPath(const std::vector<Edge>& edges, int numNodes, int start, int end) {
    std::vector<int> empty;
    if (start < 0 || start >= numNodes || end < 0 || end >= numNodes) {
        return empty;
    }
    if (hasNegativeCycle(edges, numNodes, start)) {
        return empty;
    }
    std::vector<long long> dist(numNodes, INF);
    std::vector<int> parent(numNodes, -1);
    dist[start] = 0;
    for (int pass = 0; pass < numNodes - 1; pass++) {
        bool changed = false;
        for (const Edge& edge : edges) {
            if (edge.from < 0 || edge.from >= numNodes || edge.to < 0 || edge.to >= numNodes) {
                continue;
            }
            if (dist[edge.from] != INF && dist[edge.from] + edge.weight < dist[edge.to]) {
                dist[edge.to] = dist[edge.from] + edge.weight;
                parent[edge.to] = edge.from;
                changed = true;
            }
        }
        if (!changed) {
            break;
        }
    }
    if (dist[end] == INF) {
        return empty;
    }
    std::vector<int> path;
    for (int node = end; node != -1; node = parent[node]) {
        path.push_back(node);
    }
    std::reverse(path.begin(), path.end());
    return !path.empty() && path.front() == start ? path : empty;
}

std::string timeComplexity() {
    return "O(VE)";
}

std::string spaceComplexity() {
    return "O(V)";
}

#include "question1_tests.h"

int main() {
    auto h = q1_tests::buildHarness();
    bool passed = h.run();
    std::cout << "\n";
    q1_tests::runDungeonDemo();
    return passed ? 0 : 1;
}

File: python/question1_bellman_ford.py
"""Question 1 -- Bellman-Ford: Cursed Dungeon Routes."""
from dataclasses import dataclass
from typing import List

from test_harness import TestHarness

INF = 10**18


@dataclass(frozen=True)
class Edge:
    from_node: int
    to: int
    weight: int


def shortest_distances(edges: List[Edge], num_nodes: int, start: int) -> List[int]:
    dist = [INF] * num_nodes
    if start < 0 or start >= num_nodes:
        return dist
    dist[start] = 0
    for _ in range(num_nodes - 1):
        changed = False
        for edge in edges:
            if not (0 <= edge.from_node < num_nodes and 0 <= edge.to < num_nodes):
                continue
            if dist[edge.from_node] != INF and dist[edge.from_node] + edge.weight < dist[edge.to]:
                dist[edge.to] = dist[edge.from_node] + edge.weight
                changed = True
        if not changed:
            break
    return dist


def shortest_path(edges: List[Edge], num_nodes: int, start: int, end: int) -> List[int]:
    if start < 0 or start >= num_nodes or end < 0 or end >= num_nodes:
        return []
    if has_negative_cycle(edges, num_nodes, start):
        return []
    dist = [INF] * num_nodes
    parent = [-1] * num_nodes
    dist[start] = 0
    for _ in range(num_nodes - 1):
        changed = False
        for edge in edges:
            if not (0 <= edge.from_node < num_nodes and 0 <= edge.to < num_nodes):
                continue
            if dist[edge.from_node] != INF and dist[edge.from_node] + edge.weight < dist[edge.to]:
                dist[edge.to] = dist[edge.from_node] + edge.weight
                parent[edge.to] = edge.from_node
                changed = True
        if not changed:
            break
    if dist[end] == INF:
        return []
    path = []
    node = end
    while node != -1:
        path.append(node)
        node = parent[node]
    path.reverse()
    return path if path and path[0] == start else []


def has_negative_cycle(edges: List[Edge], num_nodes: int, start: int) -> bool:
    dist = shortest_distances(edges, num_nodes, start)
    for edge in edges:
        if not (0 <= edge.from_node < num_nodes and 0 <= edge.to < num_nodes):
            continue
        if dist[edge.from_node] != INF and dist[edge.from_node] + edge.weight < dist[edge.to]:
            return True
    return False


def time_complexity() -> str:
    return "O(VE)"


def space_complexity() -> str:
    return "O(V)"


if __name__ == "__main__":
    from question1_tests import run_tests, run_dungeon_demo
    passed = run_tests()
    print()
    run_dungeon_demo()
    if not passed:
        raise SystemExit(1)
